lib/ModuleManager.cpp

```diff
@@ -466,22 +466,80 @@ void QoreModuleManager::init(bool se) {
       QoreModuleManager::addStandardModulePaths();
 }
 
+/* this internal helper function solves an issue with modules
+   on MS Windows (or maybe OSX later).
+
+   The problem with Windows is that we are distributing zip
+   files which can be extracted anywhere. But these binaries
+   had hardcoded paths for modules (/z/something/something or
+   C:/msys64/home/pvanek/src/qore/RELEASE/share/qore-modules
+   for example). These hardcoded paths are fine on Linux because
+   the language should be distributed in a form of system packages
+   (rpm, deb). But on Windows the user decides where to put
+   (extract) the zip archive content. It can be anywhere.
+
+   Of course there is QORE_MODULE_DIR env variable. But it
+   requires logout/login on Windows to be in action. Also
+   user has to set 4 paths there. And unfortunately 2 of
+   them must contain qore version in the path. So it
+   required to change env variable (and re-login) after any
+   qore upgrade. It was annoying.
+
+   So for now - if there is MODULES_RELATIVE_PATH defined
+   from cmake (WIN builds only for now) the location of
+   module(s) is constructed from the location of qore binary.
+ */
+#ifdef MODULES_RELATIVE_PATH
+#ifdef _Q_WINDOWS
+#include <windows.h>
+#endif
+#endif
+const char * module_dir_prefix(const char * path) {
+#ifdef MODULES_RELATIVE_PATH
+#ifdef _Q_WINDOWS
+    // get windows qore.exe binary
+    // Full path may be longer than MAX_PATH
+    // Let's expand it until it fits
+    std::vector<char> buff;
+    DWORD copied = 0;
+    do {
+        buff.resize(buff.size() + MAX_PATH);
+        copied = GetModuleFileName(0, &buff.at(0), buff.size());
+    } while (copied >= buff.size());
+
+    buff.resize(copied);
+    std::string prefix_path(buff.data(), buff.size());
+    size_t found_pos = prefix_path.rfind('\\');
+
+    return prefix_path.substr(0, found_pos).append("/../").append(path).c_str();
```

Comment:
after `substr.append.append` there is a temporary string variable (on stack), that will be destroyed immediately after exiting the context, then you take a pointer into its member `c_str` and you return it - after return, the temp string is destroyed and the pointer points to undefined memory - normally this should ends with deq fault (maybe sometimes with debug build it can work)

Reply:
I would return the string, it's safer
also it's a pity that API of `UniqueDirectoryList` hash push_back variant only for `DLLLOCAL bool push_back(const char* str)` even that inner implementation uses `std::set<std::string>` so the char* is the copied anyway.. Maybe introduce new API `push_back(const std::string &str)` adn `push_back(std::string &&str)`, he? :)

Reply:
no idea what to do here. Should I copy the char*? Or to return a std::string and call c_str() in the `moduleDirList.push_back()` calls?

Reply:
check it now, please